packages/backend/src/open_payments/auth/middleware.test.ts

```diff
@@ -105,83 +108,214 @@ describe('Auth Middleware', (): void => {
-      status: 403,
-      message: 'Insufficient Grant'
-    })
-    expect(introspectSpy).toHaveBeenCalledWith(token)
-    expect(next).not.toHaveBeenCalled()
-  })
+        ],
+        key
+      })
 
-  test.each`
-    limitAccount
-    ${false}
-    ${true}
-  `(
-    'sets the context grant and calls next (limitAccount: $limitAccount)',
-    async ({ limitAccount }): Promise<void> => {
-      const tokenInfo = new TokenInfo(
-        {
-          active: true,
-          clientId: uuid(),
-          grant: uuid(),
-          access: [
-            {
-              type: AccessType.IncomingPayment,
-              actions: [AccessAction.Read],
-              identifier: limitAccount ? ctx.paymentPointer.url : undefined
-            },
+      test.each`
+        type                          | action                   | description
+        ${AccessType.OutgoingPayment} | ${action}                | ${'type'}
+        ${type}                       | ${AccessAction.Complete} | ${'action'}
+      `(
+        'returns 403 for unauthorized request (conflicting $description)',
+        async ({ type, action }): Promise<void> => {
+          const tokenInfo = createTokenInfo([
             {
-              type: AccessType.OutgoingPayment,
-              actions: [AccessAction.Create, AccessAction.Read],
-              identifier: ctx.paymentPointer.url,
-              interval: 'R/2022-03-01T13:00:00Z/P1M',
-              limits: {
-                receiveAmount: {
-                  value: BigInt(500),
-                  assetCode: 'EUR',
-                  assetScale: 2
-                },
-                sendAmount: {
-                  value: BigInt(811),
-                  assetCode: 'USD',
-                  assetScale: 2
-                },
-                receiver:
-                  'https://wallet2.example/bob/incoming-payments/aa9da466-12ba-4760-9aa0-8c06061f333b'
-              }
+              type,
+              actions: [action],
+              identifier
             }
-          ]
-        },
-        key
+          ])
+          const introspectSpy = jest
+            .spyOn(authService, 'introspect')
+            .mockResolvedValueOnce(tokenInfo)
+          await expect(middleware(ctx, next)).rejects.toMatchObject({
+            status: 403,
+            message: 'Insufficient Grant'
+          })
+          expect(introspectSpy).toHaveBeenCalledWith(token)
+          expect(next).not.toHaveBeenCalled()
+        }
       )
-      const introspectSpy = jest
-        .spyOn(authService, 'introspect')
-        .mockResolvedValueOnce(tokenInfo)
 
-      await expect(middleware(ctx, next)).resolves.toBeUndefined()
-      expect(introspectSpy).toHaveBeenCalledWith(token)
-      expect(next).toHaveBeenCalled()
-      expect(ctx.grant).toEqual(tokenInfo)
+      if (identifierOption !== IdentifierOption.Conflicting) {
+        test('sets the context client info and calls next', async (): Promise<void> => {
```

Comment:
More for my own understanding, why are we ok to allow `tokenInfo` to _not_ have an identifier?

Reply:
- https://github.com/interledger/open-payments/issues/124

but then I'm not sure why it was subsequently re-required for outgoing payments :grimacing: 
- https://github.com/interledger/open-payments/pull/144

Reply:
>  identifier (string):
     A string identifier indicating a specific resource at the RS. For example, a patient identifier for a medical API or      a bank account number for a financial API.
     
Based on the description, this feels like it should be reference to specific sub resource, and _not_ just a payment pointer (of how we are currently treating it). I'm ok with keeping it optional across the board, thoughts?

Reply:
:thinking: Our own `description`s also seem to suggests it would refer to a specific (sub)resource...
https://github.com/interledger/open-payments/blob/main/openapi/schemas.yaml#L56-L59
>description: A string identifier indicating a specific resource at the RS.

Reply:
I think we do want to be able to refer to a payment pointer, especially for `create` and `list` action grants.
If/when we switch to Adrian's grant requests to the payment pointer itself, the client will no longer need to specify the payment pointer. And he's proposed just having a `paymentPointer` field in the token introspection response for the grant.